sycl/include/sycl/esimd/detail/esimd_region.hpp

```diff
@@ -0,0 +1,119 @@
+//==-------------- esimd_region.hpp - DPC++ Explicit SIMD API --------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// Region type to implement the Explicit SIMD APIs.
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <CL/sycl/detail/defines.hpp>
+#include <cstdint>
+#include <type_traits>
+#include <utility>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace intel {
+namespace gpu {
+
+// The common base type of region types.
+template <bool Is2D, typename T, int SizeY, int StrideY, int SizeX, int StrideX>
+struct region_base {
+  using element_type = T;
+  static constexpr int length = SizeX * SizeY;
+  static constexpr int Is_2D = Is2D;
+  static constexpr int Size_x = SizeX;
+  static constexpr int Stride_x = StrideX;
+  static constexpr int Size_y = SizeY;
+  static constexpr int Stride_y = StrideY;
+  static constexpr int Size_in_bytes = sizeof(T) * length;
+
+  static_assert(Size_x > 0 && Stride_x >= 0, "illegal region in x-dimension");
+  static_assert(Size_y > 0 && Stride_y >= 0, "illegal region in y-dimension");
+
+  uint16_t M_offset_y;
+  uint16_t M_offset_x;
+
+  explicit region_base() : M_offset_y(0), M_offset_x(0) {}
+
+  explicit region_base(uint16_t OffsetX) : M_offset_y(0), M_offset_x(OffsetX) {}
+
+  explicit region_base(uint16_t OffsetY, uint16_t OffsetX)
+      : M_offset_y(OffsetY), M_offset_x(OffsetX) {}
+};
+
+// A basic 1D region type.
+template <typename T, int Size, int Stride>
+using region1d_t = region_base<false, T, 1, 0, Size, Stride>;
+
+// A basic 2D region type.
+template <typename T, int SizeY, int StrideY, int SizeX, int StrideX>
+using region2d_t = region_base<true, T, SizeY, StrideY, SizeX, StrideX>;
+
+// simd_view forward declaration.
+template <typename BaseTy, typename RegionTy> class simd_view;
```

Comment:
what is a 'BaseTy'?

Reply:
BaseTy represents the type for the base SIMD object. Please refer to include/sycl/esimd/esimd_view.hpp.

Reply:
thanks, got it.